server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java

```diff
@@ -0,0 +1,93 @@
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import io.crate.data.Input;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+
+import java.util.List;
+
+import static io.crate.expression.scalar.array.ArrayArgumentValidators.ensureInnerTypeIsNotUndefined;
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+public class ArrayMaxFunction<T extends Comparable> extends Scalar<T, List<T>> {
+
+    public static final String NAME = "array_max";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                parseTypeSignature("array(E)"),
+                parseTypeSignature("E")
+            ).withTypeVariableConstraints(typeVariable("E")),
+            ArrayMaxFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private ArrayMaxFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        ensureInnerTypeIsNotUndefined(boundSignature.getArgumentDataTypes(), signature.getName().name());
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public T evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input[] args) {
+        List<T> values = (List) args[0].value();
+
+        if (values == null || values.isEmpty()) {
+            return null;
+        }
+
+        // Taking first element in order not to initialize max
+        // with type dependant TYPE.MIN_VALUE.
+        T max = values.get(0);
+
+        for (int i = 1; i < values.size(); i++) {
+            T item = values.get(i);
+            if (item.compareTo(max) > 0) {
```

Comment:
Only a few of the value instances implement `Comparable`, so this might fail at runtime for many of the supported type.

`DataType` has a `compare` function that could be used instead to support more types. Might also make sense to restrict the function registration to only match on types which we actually want to support.

Reply:
Good point, thanks.

>Restrict function registration:

Iterating over PRIMITIVE_TYPES now - similar to aggregation min/max - I believe supported type set must be equivalent.

>might fail at runtime

Updated _happy path test_ to check support of all PRIMITIVE_TYPES.

>`DataType` has a `compare` function

Done. However, not sure about usage of PRIMITIVE_TYPES now - I noticed non-primitive types with compare (GeoPoint, Regclass) - on the one hand, why not support them as they have compare, on the other hand, is it fine that aggregation `max` and `array_max` would have different argument type set?